### PR TITLE
BALANCING GROUPS: Implementation

### PR DESCRIPTION
**Pull request by ethouris**
No deps.

### CODE REVIEW

File: srtcore/core.cpp [hunk 1]
```diff
@@ -11695,11 +11717,11 @@ int CUDTGroup::send(const char* buf, int len, SRT_MSGCTRL& w_mc)
     case SRT_GTYPE_BACKUP:
         return sendBackup(buf, len, (w_mc));
 
-        /* to be implemented
-
     case SRT_GTYPE_BALANCING:
         return sendBalancing(buf, len, (w_mc));
+        //return old_sendBalancing(buf, len, r_mc);
```
Comment (maxsharabayko) [inline srtcore/core.cpp]:
```suggestion
```

File: srtcore/core.cpp [hunk 2]
```diff
@@ -11147,18 +11164,23 @@ CUDTGroup::CUDTGroup(SRT_GROUP_TYPE gtype)
     switch (gtype)
     {
     case SRT_GTYPE_BROADCAST:
+        //m_bSynchOnMsgNo = false;
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: Is this comment needed? `m_bSynchOnMsgNo`

File: srtcore/core.cpp [hunk 3]
```diff
@@ -14468,3 +14531,1224 @@ int CUDTGroup::configure(const char* str)
...
+    // if it was ever seen broken, so that it's skipped.
+    set<CUDTSocket*> broken;
+
+    // Remember them now because they will be overwritten.
+    SRT_SOCKGROUPDATA* out_grpdata = (w_mctrl.grpdata);
+    size_t out_grpdata_size = w_mctrl.grpdata_size;
+
+    size_t output_size = 0;
+
+RETRY_READING:
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: This goto label can be replaced with `while(true)` loop.

File: srtcore/core.cpp [hunk 4]
```diff
@@ -14468,3 +14531,1224 @@ int CUDTGroup::configure(const char* str)
...
+                continue;
+            }
+
+            // Check if this link was used at least once so far.
+            // If not, select it immediately.
+            if (li->load_factor == 0)
+            {
+                HLOGC(dlog.Debug, log << "linkSelect_window: ... load factor empty: SELECTING.");
+                this_link = li;
+                goto ReportLink;
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: Avoid goto labels. Make a function `CUDTGroup::gli_t reportLink(CUDTGroup::gli_t this_link)` and return its result:
```
return reportLink(this_link);
```

File: srtcore/core.cpp (hunk 2 above)
Comment (ethouris) [reply srtcore/core.cpp]: I'm still not sure. This was first done this way that the functions had common parts and this was used to distinguish synch on msg no to synch on seq no. Might be that this idea will be revived after some refax. I'd opt this to stay, at least as informative - or maybe to change it into a comment at this place.

File: srtcore/core.cpp (hunk 3 above)
Comment (ethouris) [reply srtcore/core.cpp]: Yeah, the problem is that it would likely embrace the whole function. This function is also quite long and would be nice if it can be split into smaller parts. With this goto form at least you have it in the beginning and in the end of the function and therefore it's clear enough: repeat everything from start.

File: srtcore/core.cpp (hunk 4 above)
Comment (ethouris) [reply srtcore/core.cpp]: Good idea.

File: srtcore/buffer.cpp [hunk 5]
```diff
@@ -247,7 +247,9 @@ void CSndBuffer::addBuffer(const char* data, int len, SRT_MSGCTRL& w_mctrl)
     // maximum value has been reached. Casting to int32_t to ensure the same sign
     // in comparison, although it's far from reaching the sign bit.
 
-    m_iNextMsgNo = ++MsgNo(m_iNextMsgNo);
+    int nextmsgno = ++MsgNo(m_iNextMsgNo);
```
Comment (maxsharabayko) [inline srtcore/buffer.cpp:250]: const

File: srtcore/core.cpp [hunk 6]
```diff
@@ -2111,7 +2114,7 @@ bool CUDT::createSrtHandshake(
         }
         // (this function will not fill the variables with anything, if no master is found)
 
-        int32_t storedata [GRPD__SIZE] = { id, tp, m_parent->m_IncludedIter->priority };
+        int32_t storedata [GRPD__SIZE] = { id, tp, m_parent->m_IncludedIter->weight };
```
Comment (maxsharabayko) [inline srtcore/core.cpp:2117]: const?

File: srtcore/core.cpp [hunk 7]
```diff
@@ -12515,602 +12559,599 @@ int CUDTGroup::recv(char* buf, int len, SRT_MSGCTRL& w_mc)
...
 
-RETRY_READING:
-
+    for (;;)
     {
         if (!m_bOpened || !m_bConnected)
         {
             LOGC(dlog.Error, log << boolalpha << "group/recv: ERROR opened=" << m_bOpened << " connected=" << m_bConnected);
             throw CUDTException(MJ_CONNECTION, MN_NOCONN, 0);
         }
```
Comment (maxsharabayko) [inline srtcore/core.cpp:12568]: These changes are not related to the balancing implementation and look more like a refactoring, that was requested in #1134. Having these changes in this PR complicate the reviewing process.
- [ ] Move to separate PR

File: srtcore/core.cpp [hunk 8]
```diff
@@ -14165,9 +14242,9 @@ int CUDTGroup::sendBackup(const char *buf, int len, SRT_MSGCTRL& w_mc)
         // no sendable, a new link needs to be activated anyway), and if the
         // priority has a lower number.
         if (sendable_pri.empty()
-                || (!idlers.empty() && idlers[0]->priority < *sendable_pri.begin() ))
+                || (!idlers.empty() && idlers[0]->weight < *sendable_pri.begin() ))
```
Comment (maxsharabayko) [inline srtcore/core.cpp:14245]: There are a lot of places where `priority` is renamed to `weight`. This can also be moved to a separate PR to reduce the scope of the review.
- [ ] Move `priority` renaming to a separate PR.

Comment (ethouris) [reply srtcore/core.cpp:14245]: Renaming `priority` to `weight` is a part of this PR, just like all changes that introduce new fields for the exclusive use of balancing groups.

The implementation of balancing groups with this field named "priority" doesn't make any sense. Also there were two parts of it in two different PRs, so extracting them would be purely synthetic.
